lib/iris/experimental/xarray_dataset_wrapper.py

```diff
@@ -0,0 +1,415 @@
+
+    def __len__(self):
+        return self._len
+
+    def isunlimited(self):
+        return self._unlimited
+
+
+class _Nc4AttrsMimic(_XrMimic):
+    """
+    A class mixin for a Mimic with attribute access.
+
+    I.E. shared by variables and datasets.
+
+    """
+
+    def ncattrs(self):
+        return self._xr.attrs.keys()  # Probably do *not* need/expect a list ?
+
+    def getncattr(self, attr_name):
+        if attr_name in self._xr.attrs:
+            result = self._xr.attrs[attr_name]
+        else:
+            raise AttributeError()
+        return result
+
+    def __getattr__(self, attr_name):
+        return self.getncattr(attr_name)
+
+    #
+    # writing
+    #
+    def setncattr(self, attr_name, value):
+        if isinstance(value, bytes):
+            value = value.decode()
+        self._xr.attrs[attr_name] = value
+
+    # NOTE: not currently supporting ".my_attribute = value" type access.
+    # def __setattr__(self, attr_name, value):
+    #     self.setncattr(attr_name, value)
+
+
+class VariableMimic(_Nc4AttrsMimic):
+    """
+    A Variable object mimic wrapper.
+
+    Variable additional properties:
+        dimensions, dtype, data (+ attributes, parent-group)
+        shape, size, ndim
+
+    """
+
+    @property
+    def dtype(self):
+        return self._xr.dtype
+
+    def chunking(self):
+        return None
+
+    @property
+    def datatype(self):
+        return self.dtype
+
+    @property
+    def dimensions(self):
+        return self._xr.dims
+
+    def __getitem__(self, keys):
+        if self.ndim == 0:
+            return self._xr.data
+        else:
+            return self._xr[keys].data
+
+    @property
+    def shape(self):
+        return self._xr.shape
+
+    @property
+    def ndim(self):
+        return self._xr.ndim
+
+    @property
+    def size(self):
+        return self._xr.size
+
+    #
+    # writing
+    #
+    def __setitem__(self, keys, data):
+        self._xr[keys] = data
+
+
+class DatasetMimic(_Nc4AttrsMimic):
+    """
+    An object mimicking an netCDF4.Dataset, wrapping an xarray.Dataset.
+
+    """
+
+    def __init__(self, xrds=None):
+        """
+        Create a Dataset mimic, which provides a bridge between the
+        :class:`netcdf.Dataset` access API and data in the form of an
+        :class:`xarray.Dataset`.
+
+        Parameters
+        ----------
+        xrds : :class:`xr.Dataset`, optional
+            If provided, create a DatasetMimic representing the xarray data.
+            If None, initialise empty.
+            In either case, the result can be read or written like a
+            :class:`netcdf.Dataset`.  Or, an xarray equivalent can be
+            regenerated with the :meth:`to_xarray_dataset` method.
+
+        Notes
+        -----
+        Only a limited subset of the :mod:`netCDF4` APIs are currently
+        supported : just enough to allow Iris to read and write xarray datasets
+        in place of netcdf files.
```

Comment:
Again, this might be where a runtime protocol could help document which parts of the api are implemented so far.